[DataPipe] Update docstring for functional form of DataPipes

weiji14 opened the pull request:
Copy the docstring from IterDataPipe and MapDataPipe classes to their functional form. Done using [`functools.update_wrapper`](https://docs.python.org/3/library/functools.html#functools.update_wrapper), xref https://stackoverflow.com/questions/6394511/python-functools-wraps-equivalent-for-classes.

See also parallel change to `.pyi` stub files at https://github.com/pytorch/pytorch/pull/100503

Fixes https://github.com/pytorch/data/issues/792 and https://github.com/weiji14/zen3geo/issues/69.

weiji14 committed:
Update docstring for functional form of IterDataPipes

Copy the docstring from IterDataPipe classes to their functional form. Xref https://github.com/pytorch/data/issues/792.

weiji14 committed:
Fix typo

ejguan reviewed (commented):
Thanks, changes look good. Is there a way for us to test this change?

NivekT reviewed (commented):
I just tried to see if the inline doc shows up in PyCharm with this change:
```python
dp = IterableWrapper(range(10)).shuffle()
```

And it doesn't seem to.  Am I missing something? Perhaps changes to `datapipe.fyi` is necessary? I believe it is currently rendering whatever is in `datapipe.fyi`.

<img width="605" alt="Screenshot 2023-05-02 at 2 24 49 PM" src="https://user-images.githubusercontent.com/4935152/235752780-cf290377-4f38-4227-bc7e-2156a2eddc9c.png">


Not sure if this is related, but it is worth noting this works even before this change:
```python
dp = IterableWrapper(range(10)).shuffle()
print(dp.__doc__)  # This prints the right string even before this change
```

Let me know if none of the above is relevant to this PR.

weiji14 commented:
> And it doesn't seem to. Am I missing something? Perhaps changes to `datapipe.fyi` is necessary? I believe it is currently rendering whatever is in `datapipe.fyi`.

How do I regenerate the `torch/utils/data/datapipes/datapipe.pyi.in` file? To be honest, I haven't quite figured out how to check this locally yet, it's my first time working on the pytorch codebase :sweat_smile:

NivekT commented:
Re-installing PyTorch locally (`python setup.py develop`) will regenerates that file. Note that file is generated from `datapipe.pyi.in` and `gen_pyi.py` in `torch.utils.data.datapipes`.

weiji14 commented:
> Re-installing PyTorch locally (`python setup.py develop`) will regenerates that file. Note that file is generated from `datapipe.pyi.in` and `gen_pyi.py` in `torch.utils.data.datapipes`.

Ok, let me try that. It's taking a bit of time to initialize the third party submodules.

weiji14 commented:
Hmm, I'm wondering if some change is needed on the torchdata repo too as mentioned in https://github.com/pytorch/pytorch/pull/100503#issue-1693166064? I've regenerated the `torch/utils/data/datapipes/datapipe.pyi` file using the patched `torch/utils/data/datapipes/gen_pyi.py` script in #100503 which matches https://gist.github.com/NivekT/95095f14da85a837a0727a19a5ba367c, but it seems like `help(dp.batch)` still returns the partial docstring?

NivekT commented:
@weiji14 I have only checked the doc string that was shown in an IDE (which becomes correct with #100503), I didn't try `help(dp)`. 

Does this PR allows `help(dp.batch)` to render correctly (i.e. equivalent to `help(Batcher)`? If that is the case, we can land both PRs.

weiji14 committed:
Add regression test to check correctness of DataPipe functional form

Ensure that the functional form of IterDataPipe and MapDataPipe contains the proper docstring copied over from the class form.

weiji14 committed:
Remove unzip test

UnZipperMapDataPipe is only in https://github.com/pytorch/data/blob/v0.6.0/torchdata/datapipes/map/util/unzipper.py#L16-L17 and not pytorch/pytorch repo.

weiji14 reviewed (commented):
> @weiji14 I have only checked the doc string that was shown in an IDE (which becomes correct with #100503), I didn't try `help(dp)`.
> 
> Does this PR allows `help(dp.batch)` to render correctly (i.e. equivalent to `help(Batcher)`? If that is the case, we can land both PRs.

Yes, I think this PR makes `help(dp.batch)` equivalent to `help(Batcher)`. Have added a unit test in 452f254189d3778b9c2e475f4d3cb5548969237e to check for this (though it uses `pydoc.render` following https://stackoverflow.com/questions/7123660/is-there-an-option-to-print-the-output-of-help). From what I can tell, the `.pyi` stub file updates at #100503 is only for the type hints according to https://www.jetbrains.com/help/pycharm/stubs.html?

NivekT commented on test/test_datapipe.py:
```diff
@@ -831,6 +832,35 @@ def _fn3(x):
+            "map",
+            "mux",
+            "read_from_stream",
+            # "sampler",
```
I think we left it out for a reason, we can add the functional form for `sampler` later if there is demand for me

NivekT reviewed (approved):
LGTM! Thanks for adding this!

NivekT commented:
@pytorchbot merge -r

NivekT reviewed (changes requested):
@weiji14 Can you have a look at the [failing CI tests](https://github.com/pytorch/pytorch/actions/runs/4929875273/jobs/8810511389)? They seem relevant

weiji14 commented on test/test_datapipe.py:
```diff
@@ -831,6 +832,35 @@ def _fn3(x):
+            "zip",
+        ]:
+            docstring = pydoc.render_doc(thing=getattr(input_dp, dp_funcname))
+            assert f"(functional name: ``{dp_funcname}``)" in docstring
```
> @weiji14 Can you have a look at the [failing CI tests](https://github.com/pytorch/pytorch/actions/runs/4929875273/jobs/8810511389)? They seem relevant

Yes I see the error, though it seems to be only for certain combinations of the build matrix which is weird. Traceback at https://github.com/pytorch/pytorch/actions/runs/4929875273/jobs/8810511389#step:13:628:

```
==================================== RERUNS ====================================
__________________ TestFunctionalIterDataPipe.test_docstring ___________________
Traceback (most recent call last):
  File "test_datapipe.py", line 860, in test_docstring
    assert f"(functional name: ``{dp_funcname}``)" in docstring
AssertionError
__________________ TestFunctionalIterDataPipe.test_docstring ___________________
Traceback (most recent call last):
  File "test_datapipe.py", line 860, in test_docstring
    assert f"(functional name: ``{dp_funcname}``)" in docstring
AssertionError
=================================== FAILURES ===================================
__________________ TestFunctionalIterDataPipe.test_docstring ___________________
Traceback (most recent call last):
  File "test_datapipe.py", line 860, in test_docstring
    assert f"(functional name: ``{dp_funcname}``)" in docstring
AssertionError
- generated xml file: /var/lib/jenkins/workspace/test/test-reports/python-pytest/test_datapipe/test_datapipe-17e8ddb9c33a3903.xml -
=========================== short test summary info ============================
FAILED [0.0028s] test_datapipe.py::TestFunctionalIterDataPipe::test_docstring - AssertionError
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
=============== 1 failed, 22 passed, 7 skipped, 2 rerun in 0.45s ===============
Got exit code 1, retrying (retries left=2)
Test results will be stored in test-reports/python-pytest/test_datapipe/test_datapipe-0b2a1e96f86b787b.xml
```
Logs aren't very helpful, let me try and track down the issue.

weiji14 replied:
Merging in the changes from #100503 to see if it helps. I can't seem to reproduce this locally on my setup, and not quite sure what the shard/num_shards config means in the `linux-bionic-py3_8-clang9-build` CI build matrix at https://github.com/pytorch/pytorch/blob/daed3bf8f9d10367ae3a34d8ea6ca2b594f9afe2/.github/workflows/pull.yml#L124C1-L139

weiji14 replied:
Have tried using `pydoc.render(..., forceload=True)` at c21eaa557d21a751ab78c3b93ba02a38a1113f7e, but somehow the docstring still isn't updated on some build matrix combinations. Getting an output like so from https://github.com/pytorch/pytorch/actions/runs/4955538159/jobs/8877794743?pr=100446#step:14:1087:

```python-traceback
  __________________ TestFunctionalIterDataPipe.test_docstring ___________________
  Traceback (most recent call last):
    File "test_datapipe.py", line 863, in test_docstring
      assert f"(functional name: ``{dp_funcname}``)" in docstring
  AssertionError
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "test_datapipe.py", line 868, in test_docstring
      raise ValueError(dp_funcname, "IterDataPipe docstring incorrect")
  ValueError: ('batch', 'IterDataPipe docstring incorrect')
  ----------------------------- Captured stdout call -----------------------------
  ***Begin docstring for batch
  Python Library Documentation: partial in module torch.utils.data.datapipes.iter.grouping object
  
  class ppaarrttiiaall(builtins.object)
   |  partial(func, *args, **keywords) - new function with partial application
   |  of the given arguments and keywords.
   |
   |  Methods defined here:
   |
   |  ____ccaallll____(self, /, *args, **kwargs)
   |      Call self as a function.
   |
   |  ____ddeellaattttrr____(self, name, /)
   |      Implement delattr(self, name).
   |
   |  ____ggeettaattttrriibbuuttee____(self, name, /)
   |      Return getattr(self, name).
   |
   |  ____rreedduuccee____(...)
   |      Helper for pickle.
   |
   |  ____rreepprr____(self, /)
   |      Return repr(self).
   |
   |  ____sseettaattttrr____(self, name, value, /)
   |      Implement setattr(self, name, value).
   |
   |  ____sseettssttaattee____(...)
   |
   |  ----------------------------------------------------------------------
   |  Static methods defined here:
   |
   |  ____nneeww____(*args, **kwargs) from builtins.type
   |      Create and return a new object.  See help(type) for accurate signature.
   |
   |  ----------------------------------------------------------------------
   |  Data descriptors defined here:
   |
   |  ____ddiicctt____
   |
   |  aarrggss
   |      tuple of arguments to future partial calls
   |
   |  ffuunncc
   |      function object to use in future partial calls
   |
   |  kkeeyywwoorrddss
   |      dictionary of keyword arguments to future partial calls
  
  ***End docstring for batch
  __________________ TestFunctionalIterDataPipe.test_docstring ___________________
  Traceback (most recent call last):
    File "test_datapipe.py", line 863, in test_docstring
      assert f"(functional name: ``{dp_funcname}``)" in docstring
  AssertionError
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "test_datapipe.py", line 868, in test_docstring
      raise ValueError(dp_funcname, "IterDataPipe docstring incorrect")
  ValueError: ('batch', 'IterDataPipe docstring incorrect')
```

Not sure why the docstring output is repeated in some parts :confused: I could xfail those tests for now and investigate them later?

weiji14 committed:
Debug failing test_docstring with print statements

weiji14 committed:
Delete extra space

weiji14 committed:
Tweak debugging print statements

weiji14 committed:
Set forceload=True in pydoc.render_doc

Force reload the module from disk to see if it helps with the test.

NivekT reviewed (commented):
@weiji14 It looks like it is failing on Python 3.8, can you try to run it in a local Linux environment with 3.8? I think that should help you reproduce the issue.

weiji14 committed:
Make test_docstring compatible with Python 3.8

weiji14 committed:
Remove debugging assert statements

Reverts 426e244fc66e47ad12085447969cc1754f3fe563

weiji14 commented:
> @weiji14 It looks like it is failing on Python 3.8, can you try to run it in a local Linux environment with 3.8? I think that should help you reproduce the issue.

Thanks for the tip! Apparently `pydoc.render_doc` works differently on Python 3.8 and Python 3.9+, see https://docs.python.org/3/whatsnew/3.9.html#pydoc and https://github.com/python/cpython/issues/84438. The output of `help(dp.batch)` is still the same though (just that `help()` renders things to stdout). I've made a commit at 52be9e1c83767a5ec1dfd168752ddbf46a688f9c that should make the unit test work for both Python 3.8 and Python 3.9+ :crossed_fingers:

weiji14 committed:
Apply patch for Python 3.8 on TestFunctionalMapDataPipe.test_docstring

Same as 52be9e1c83767a5ec1dfd168752ddbf46a688f9c  but for MapDataPipe

NivekT reviewed (approved):
LGTM! Thanks!

NivekT commented:
@pytorchbot merge -r
